README.md
Participation credit
====================

[![Build Status](https://travis-ci.org/mudrd8mz/moodle-workshopeval_credit.svg?branch=master)](https://travis-ci.org/mudrd8mz/moodle-workshopeval_credit)

Simple grading evaluation method for Moodle Workshop module that gives credit
to peer-reviewers just for their participation in the activity without
comparing their assessment with the others.

There are three modes of how grades for assessment are calculated by this method.

* All or nothing - The reviewer must assess all allocated submissions in order
  to obtain the maximum grade; otherwise they receive a grade of zero.
* Proportional - The grade obtained is proportional to the number of
  assessments. If all allocated submissions are assessed, the reviewer will
  obtain the maximum grade; if half of the allocated submissions are assessed,
  the reviewer will obtain 50% of the maximum grade.
* At least one - The reviewer must assess at least one allocated submission in
  order to obtain the maximum grade.


Installation
------------

Unzip the plugin into the folder mod/workshop/eval/credit/ on your Moodle site
and visit the main administration page.


Usage
-----

When the workshop activity is in the grading evaluation phase, choose the evaluation
method "Participation credit". Then select the required evaluation mode and click
"Re-calculate grades".


CHANGES.md
### 3.1.0 ###

* Implement the Privacy API, making the plugin GDPR-friendly.


tests/lib_test.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Provides thes {@link workshopeval_credit_testcase} class.
 *
 * @package     workshopeval_credit
 * @category    test
 * @copyright   2013 David Mudrak <david@moodle.com>
 * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die();

global $CFG;
require_once($CFG->dirroot.'/mod/workshop/locallib.php');
require_once($CFG->dirroot.'/mod/workshop/eval/credit/lib.php');


/**
 * Unit tests for the 'Participation credit' workshop evaluation method.
 *
 * @copyright 2013 David Mudrak <david@moodle.com>
 * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class workshopeval_credit_testcase extends advanced_testcase {

    /** @var workshop instance emulation */
    protected $workshop;

    /** @var testable_workshop_credit_evaluation */
    protected $eval;

    /**
     * Prepare the test environment.
     */
    protected function setUp() {
        global $CFG;

        parent::setUp();

        if ($CFG->version >= 2014051200) {
            // Since Moodle 2.7 we have data generators available.
            $this->setAdminUser();
            $course = $this->getDataGenerator()->create_course();
            $workshop = $this->getDataGenerator()->create_module('workshop', array('course' => $course, 'evaluation' => 'best'));
            $cm = get_coursemodule_from_instance('workshop', $workshop->id, $course->id, false, MUST_EXIST);
            $this->workshop = new workshop($workshop, $cm, $course);

        } else {
            $cm = new stdClass();
            $course = new stdClass();
            $context = new stdClass();
            $workshop = (object)array('id' => 22, 'evaluation' => 'best');
            $this->workshop = new workshop($workshop, $cm, $course, $context);
        }

        $this->eval = new testable_workshop_credit_evaluation($this->workshop);
    }

    /**
     * Shut down the test environment.
     */
    protected function tearDown() {
        $this->workshop = null;
        $this->eval = null;
        parent::tearDown();
    }

    /**
     * With no assessments, no grades are to be calculated.
     */
    public function test_calculate_assessment_grades_empty() {
        $this->resetAfterTest(true);

        $this->assertEquals(array(), $this->eval->calculate_assessment_grades(array(), 'all'));
        $this->assertEquals(array(), $this->eval->calculate_assessment_grades(array(), 'proportional'));
        $this->assertEquals(array(), $this->eval->calculate_assessment_grades(array(), 'one'));
    }

    /**
     * Test the actual grading grade calculations.
     */
    public function test_calculate_assessment_grades() {
        $this->resetAfterTest(true);

        // Prepare a structure as returned by {@link workshop_credit_evaluation::make_assessments_map()}.
        $assessments = array(
            4 => array(
                102 => null,
                103 => null,
                104 => null,
            ),
            6 => array(
                104 => null,
                105 => 100.00000,
                106 => 12.12345,
                107 => 0.00000,
            ),
            8 => array(
                106 => null,
                107 => 50.00000,
                108 => null,
            ),
            23 => array(
                108 => 100.00000,
                109 => 100.00000,
                110 => 67.65356,
            ),
            0 => array(),
            -1 => null,
        );

        // Test that non-sense data are not returned.
        $grades = $this->eval->calculate_assessment_grades($assessments, 'one');
        $this->assertFalse(array_key_exists(0, $grades));
        $this->assertFalse(array_key_exists(-1, $grades));

        // Test 'All or nothing' mode.
        $grades = $this->eval->calculate_assessment_grades($assessments, 'all');
        $this->assertSame(0.00000, $grades[4]);
        $this->assertSame(0.00000, $grades[6]);
        $this->assertSame(0.00000, $grades[8]);
        $this->assertSame(100.00000, $grades[23]);

        // Test 'At least one' mode.
        $grades = $this->eval->calculate_assessment_grades($assessments, 'one');
        $this->assertSame(0.00000, $grades[4]);
        $this->assertSame(100.00000, $grades[6]);
        $this->assertSame(100.00000, $grades[8]);
        $this->assertSame(100.00000, $grades[23]);

        // Test 'Proportional' mode.
        $grades = $this->eval->calculate_assessment_grades($assessments, 'proportional');
        $this->assertSame(0.00000, $grades[4]);
        $this->assertSame(75.00000, $grades[6]);
        $this->assertSame(33.33333, $grades[8]);
        $this->assertSame(100.00000, $grades[23]);
    }
}


/**
 * Provides access to protected methods we want to test.
 *
 * @copyright 2013 David Mudrak <david@moodle.com>
 * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class testable_workshop_credit_evaluation extends workshop_credit_evaluation {

    /**
     * Expose parent's protected method so that it can be tested by calling it directly.
     *
     * @param array $assessments
     * @param string $mode
     * @return array
     */
    public function calculate_assessment_grades(array $assessments, $mode) { // @codingStandardsIgnoreLine overriding expected
        return parent::calculate_assessment_grades($assessments, $mode);
    }
}


version.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Defines the version of the subplugin
 *
 * @package    workshopeval_credit
 * @copyright  2013 David Mudrak <david@moodle.com>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die();

$plugin->component  = 'workshopeval_credit';
$plugin->release    = '3.1.0';
$plugin->version    = 2018061800;
$plugin->requires   = 2014051200;
$plugin->maturity   = MATURITY_STABLE;


settings.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Site-level plugin configuration.
 *
 * @package     workshopeval_credit
 * @copyright   2013 David Mudrak <david@moodle.com>
 * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die();

$options = array(
    'all' => get_string('modeall', 'workshopeval_credit'),
    'proportional' => get_string('modeproportional', 'workshopeval_credit'),
    'one' => get_string('modeone', 'workshopeval_credit'),
);

$settings->add(new admin_setting_configselect('workshopeval_credit/defaultmode', get_string('mode', 'workshopeval_credit'),
    get_string('mode_desc', 'workshopeval_credit'), 'proportional', $options));


classes/privacy/provider.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Defines {@link \workshopeval_credit\privacy\provider} class.
 *
 * @package     workshopeval_credit
 * @category    privacy
 * @copyright   2018 David Mudrák <david@moodle.com>
 * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

namespace workshopeval_credit\privacy;

defined('MOODLE_INTERNAL') || die();

/**
 * Privacy API implementation for the Participation credit plugin.
 *
 * @copyright  2018 David Mudrák <david@moodle.com>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class provider implements \core_privacy\local\metadata\null_provider {

    use \core_privacy\local\legacy_polyfill;

    /**
     * Returns stringid of a text explaining that this plugin stores no personal data.
     *
     * @return string
     */
    public static function _get_reason() {
        return 'privacy:metadata';
    }
}


lang/en/workshopeval_credit.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Strings for the Participation credit workshop evaluator
 *
 * @package    workshopeval_credit
 * @category   string
 * @copyright  2013 David Mudrak <david@moodle.com>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

$string['mode'] = 'Evaluation mode';
$string['mode_desc'] = 'Default grading evaluation mode used by the _Participation credit_ method.';
$string['mode_help'] = 'The mode determines how grades for assessment are calculated.

* All or nothing - The reviewer must assess all allocated submissions in order to obtain the maximum grade; otherwise they receive a grade of zero.
* Proportional - The grade obtained is proportional to the number of assessments. If all allocated submissions are assessed, the reviewer will obtain the maximum grade; if half of the allocated submissions are assessed, the reviewer will obtain 50% of the maximum grade.
* At least one - The reviewer must assess at least one allocated submission in order to obtain the maximum grade.';
$string['modeall'] = 'All or nothing';
$string['modeone'] = 'At least one';
$string['modeproportional'] = 'Proportional';
$string['pluginname'] = 'Participation credit';
$string['privacy:metadata'] = 'Participation credit does not store any personal data';


lib.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Contains logic class and interface for the grading evaluation plugin "Participation credit".
 *
 * @package    workshopeval_credit
 * @copyright  2013 David Mudrak <david@moodle.com>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die();

require_once(dirname(dirname(__FILE__)) . '/lib.php');
require_once($CFG->libdir . '/gradelib.php');
require_once($CFG->dirroot . '/lib/formslib.php');

/**
 * Defines the computation logic of the grading evaluation subplugin
 *
 * @copyright 2013 David Mudrak <david@moodle.com>
 * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class workshop_credit_evaluation extends workshop_evaluation {

    /** @var workshop the parent workshop instance */
    protected $workshop;

    /** @var recently used settings in this workshop */
    protected $settings;

    /**
     * Constructor
     *
     * @param workshop $workshop The workshop api instance
     * @return void
     */
    public function __construct(workshop $workshop) {
        $this->workshop = $workshop;
        $this->load_settings();
    }

    /**
     * Calculates the grades for assessment and updates 'gradinggrade' fields in 'workshop_assessments' table
     *
     * @param stdClass $settings settings for this round of evaluation
     * @param null|int|array $restrict if null, update all reviewers, otherwise update just grades for the given reviewers(s)
     *
     * @return void
     */
    public function update_grading_grades(stdClass $settings, $restrict=null) {
        global $DB;

        $this->save_settings($settings);

        $assessments = $this->make_assessments_map($restrict);
        $grades = $this->calculate_assessment_grades($assessments, $settings->mode);

        foreach ($assessments as $reviewerid => $submissiongrades) {
            foreach ($submissiongrades as $submissionid => $notused) {
                $DB->set_field('workshop_assessments', 'gradinggrade', $grades[$reviewerid],
                    array('submissionid' => $submissionid, 'reviewerid' => $reviewerid));
            }
        }
    }

    /**
     * Returns an instance of the form to define evaluation settings.
     *
     * @param moodle_url $actionurl The URL to submit the settings form to.
     * @return workshop_credit_evaluation_settings_form
     */
    public function get_settings_form(moodle_url $actionurl=null) {

        $customdata['workshop'] = $this->workshop;
        $customdata['current'] = $this->settings;
        $attributes = array('class' => 'evalsettingsform credit');

        return new workshop_credit_evaluation_settings_form($actionurl, $customdata, 'post', '', $attributes);
    }

    /**
     * Delete all data related to a given workshop module instance
     *
     * @see workshop_delete_instance()
     * @param int $workshopid id of the workshop module instance being deleted
     * @return void
     */
    public static function delete_instance($workshopid) {
        global $DB;

        $DB->delete_records('workshopeval_credit_settings', array('workshopid' => $workshopid));
    }

    // Internal methods start here.

    /**
     * Loads the evaluation settings to be used in this workshop.
     *
     * @return stdClass
     */
    protected function load_settings() {
        global $DB;

        $this->settings = new stdClass();

        $saved = $DB->get_record('workshopeval_credit_settings', array('workshopid' => $this->workshop->id));

        if (isset($saved->lastmode)) {
            $this->settings->mode = $saved->lastmode;
        }

        return $this->settings;
    }

    /**
     * Saves the evaluation settings used in this workshop.
     *
     * @param stdClass $current current settings to be stored
     */
    protected function save_settings(stdClass $current) {
        global $DB;

        if (!isset($this->settings->mode)) {
            $record = new stdClass();
            $record->workshopid = $this->workshop->id;
            // The 'mode' is a reserved XMLDB keyword.
            $record->lastmode = $current->mode;
            $DB->insert_record('workshopeval_credit_settings', $record);

        } else if ($this->settings->mode != $current->mode) {
            $DB->set_field('workshopeval_credit_settings', 'lastmode', $current->mode,
                    array('workshopid' => $this->workshop->id));
        }
    }

    /**
     * Prepares an overview of assessments allocated to reviewers in this workshop.
     *
     * Returned structure is a two-dimensional array with keys [reviewerid][submissionid].
     * The item in this array is the grade given by the reviewer for the submission. The grade
     * is either null (not graded yet) or a decimal number from 0.00000 to 100.00000.
     *
     * @param null|int|array $restrict if null, returns info for all reviewers, just given reviewers(s)
     * @return array
     */
    protected function make_assessments_map($restrict) {
        global $DB;

        $assessments = array();

        $sql = "SELECT a.submissionid, a.reviewerid, a.grade
                  FROM {workshop_submissions} s
                  JOIN {workshop_assessments} a ON a.submissionid = s.id
                 WHERE s.workshopid = :workshopid AND s.example = 0";

        $params = array('workshopid' => $this->workshop->id);

        // If the $restrict is null, then update all reviewers. Otherwise add conditions.
        if (!is_null($restrict)) {
            if (!empty($restrict)) {
                list($usql, $uparams) = $DB->get_in_or_equal($restrict, SQL_PARAMS_NAMED);
                $sql .= " AND a.reviewerid $usql";
                $params = array_merge($params, $uparams);
            } else {
                throw new coding_exception('Empty value is not a valid $restrict parameter value.');
            }
        }

        $rs = $DB->get_recordset_sql($sql, $params);

        foreach ($rs as $r) {
            $assessments[$r->reviewerid][$r->submissionid] = $r->grade;
        }

        $rs->close();

        return $assessments;
    }

    /**
     * Calculates grades for assessments.
     *
     * @param array $assessments the structure returned by {@link self::make_assessments_map()}
     * @param string $mode mode evaluation (all, proportional, one)
     * @return array of (int)reviewerid => (float)gradinggrade
     */
    protected function calculate_assessment_grades(array $assessments, $mode) {

        $allocated = array(); // Number of allocated assessments per reviewer.
        $finished = array(); // Number of actually graded assessments per reviewer.
        $grades = array(); // Suggested grade for assessment (0.00000 - 100.00000).

        foreach ($assessments as $reviewerid => $submissiongrades) {
            if (!is_array($submissiongrades) or empty($submissiongrades)) {
                continue;
            }
            if (!isset($allocated[$reviewerid])) {
                $allocated[$reviewerid] = count($submissiongrades);
            }
            if (!isset($finished[$reviewerid])) {
                $finished[$reviewerid] = 0;
            }
            foreach ($submissiongrades as $submissionid => $grade) {
                if (!is_null($grade)) {
                    $finished[$reviewerid]++;
                }
            }
        }

        foreach (array_keys($allocated) as $reviewerid) {
            $grades[$reviewerid] = null;

            if ($mode === 'all') {
                if ($allocated[$reviewerid] == $finished[$reviewerid]) {
                    $grades[$reviewerid] = 100.00000;
                } else {
                    $grades[$reviewerid] = 0.00000;
                }

            } else if ($mode === 'one') {
                if ($finished[$reviewerid] > 0.00000) {
                    $grades[$reviewerid] = 100.00000;
                } else {
                    $grades[$reviewerid] = 0.00000;
                }

            } else if ($mode === 'proportional') {
                $grades[$reviewerid] = round(100 * $finished[$reviewerid] / $allocated[$reviewerid], 5);

            } else {
                throw new coding_exception('Unknown evaluation mode', $mode);
            }
        }

        return $grades;
    }
}


/**
 * Represents the settings form for this plugin.
 *
 * @copyright 2013 David Mudrak <david@moodle.com>
 * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class workshop_credit_evaluation_settings_form extends workshop_evaluation_settings_form {

    /**
     * Defines specific fields for this evaluation method.
     */
    protected function definition_sub() {
        $mform = $this->_form;

        $pluginconfig = get_config('workshopeval_credit');
        $current = $this->_customdata['current'];

        $options = array(
            'all' => get_string('modeall', 'workshopeval_credit'),
            'proportional' => get_string('modeproportional', 'workshopeval_credit'),
            'one' => get_string('modeone', 'workshopeval_credit'),
        );
        $mform->addElement('select', 'mode', get_string('mode', 'workshopeval_credit'), $options);
        $mform->addHelpButton('mode', 'mode', 'workshopeval_credit');
        $mform->setDefault('mode', isset($pluginconfig->defaultmode) ? $pluginconfig->defaultmode : 'proportional');

        $this->set_data($current);
    }
}
